File: src/Evaluate.cpp
#include "nebula/Evaluate.hpp"
#include "nebula/Values.hpp"

namespace nebula
{

int Evaluate::evaluate(const Board& board)
{
    int score = 0;

    // one calculation
    double phase = phase_of_game(board);
    
    score += material(board, phase);
    score += castling_bonus(board, phase);
    score += pawn_structure(board, phase);
    
    return (board.turn() == Color::White) ? score : -score;
}

int Evaluate::material(const Board& board, double phase)
{
    int mat = 0;
    int opening_score = 0;
    int endgame_score = 0;

    for(int piece = 0; piece < Board::num_piece_types; ++piece)
    {
        PieceType pt = static_cast<PieceType>(piece);
        
        // white pieces
        uint64_t white_pieces = board.pieces(Color::White, pt);
        while(white_pieces)
        {
            int sq = __builtin_ctzll(white_pieces);
            mat += Values::material_value[piece];
            opening_score += Values::pst[piece][sq];
            endgame_score += Values::pst_endgame[piece][sq];
            white_pieces &= white_pieces - 1;
        }
        
        // black pieces
        uint64_t black_pieces = board.pieces(Color::Black, pt);
        while(black_pieces)
        {
            int sq = __builtin_ctzll(black_pieces);
            mat -= Values::material_value[piece];
            // mirror PST scores vertically
            opening_score -= Values::pst[piece][sq ^ 56];
            endgame_score -= Values::pst[piece][sq ^ 56];
            black_pieces &= black_pieces - 1;
        }
    }

    int blended_pst = static_cast<int>(opening_score * phase + endgame_score * (1.0 - phase));

    return mat + blended_pst;
}

int Evaluate::castling_bonus(const Board& board, double phase)
{
    int bonus = 0;

    // castling rights bonus
    if(board.castling() & (board.castle_K | board.castle_Q))
        bonus += Values::castle_rights_bonus;
    if(board.castling() & (board.castle_k | board.castle_q))
        bonus -= Values::castle_rights_bonus;

    // detect castled position
    int wk = board.king_sq(Color::White);
    if(wk == 6 || wk == 2)
        bonus += Values::castled_position_bonus;
    int bk = board.king_sq(Color::Black);
    if(bk == 62 || bk == 58)
        bonus -= Values::castled_position_bonus;

    // blend by opening phase
    bonus = static_cast<int>(bonus * phase);

    return bonus;
}

int Evaluate::pawn_structure(const Board& board, double phase)
{
    int score = 0;
    
    // weaknesses for each color
    score += analyze_pawn_weaknesses(board, Color::White, phase);
    score -= analyze_pawn_weaknesses(board, Color::Black, phase);
    
    // passed pawns for each color
    score += analyze_passed_pawns(board, Color::White, phase);
    score -= analyze_passed_pawns(board, Color::Black, phase);
    
    return score;
}

int Evaluate::analyze_pawn_weaknesses(const Board& board, Color color, double phase)
{
    int penalty = 0;
    uint64_t pawns = board.pieces(color, PieceType::Pawn);
    
    // pawns per file
    std::array<int, 8> file_counts = { 0 };
    uint64_t temp_pawns = pawns;

    while(temp_pawns)
    {
        int sq = __builtin_ctzll(temp_pawns);

        ++file_counts[sq & 7];

        temp_pawns &= temp_pawns - 1;
    }
    
    // score for each pawn
    temp_pawns = pawns;
    while(temp_pawns)
    {
        int sq = __builtin_ctzll(temp_pawns);

        int file = sq & 7;
        
        // isolated pawn penalty
        if(is_isolated_pawn(board, color, file))
        {
            int isolated_penalty = Values::isolated_pawn_penalty;

            // even worse in endgame
            isolated_penalty = static_cast<int>(isolated_penalty * (1.0 + (1.0 - phase) * 0.5));
            
            penalty += isolated_penalty;
        }
        
        // doubled pawn penalty
        if(file_counts[file] > 1)
            penalty += Values::doubled_pawn_penalty;
        
        // backward pawn penalty
        if(is_backward_pawn(board, color, sq))
            penalty += Values::backward_pawn_penalty;
        
        temp_pawns &= temp_pawns - 1;
    }
    
    return -penalty; //  these are penalties
}

int Evaluate::analyze_passed_pawns(const Board& board, Color color, double phase)
{
    int bonus = 0;
    uint64_t pawns = board.pieces(color, PieceType::Pawn);
    
    while(pawns)
    {
        int sq = __builtin_ctzll(pawns);
        
        if(is_passed_pawn(board, color, sq)) {
            int rank = (color == Color::White) ? (sq >> 3) : (7 - (sq >> 3));
            bonus += passed_pawn_value(rank, phase);
            
            // additional bonuses for advanced passed pawns
            if(rank >= 5)
            {
                // check if pawn is connected to other pawns
                uint64_t adjacent_files = 0;
                int file = sq & 7;
                if(file > 0)
                    adjacent_files |= board.pieces(color, PieceType::Pawn) & (0x0101010101010101ULL << (file - 1));
                if(file < 7)
                    adjacent_files |= board.pieces(color, PieceType::Pawn) & (0x0101010101010101ULL << (file + 1));
                
                if(adjacent_files)
                    bonus += Values::connected_passed_pawn_bonus;
                
                // check if pawn is protected by another pawn
                uint64_t protection_mask = (color == Color::White) ? ((1ULL << (sq - 7)) | (1ULL << (sq - 9))) : ((1ULL << (sq + 7)) | (1ULL << (sq + 9)));
                
                if(board.pieces(color, PieceType::Pawn) & protection_mask)
                    bonus += Values::protected_passed_pawn_bonus;
            }
        }
        
        pawns &= pawns - 1;
    }
    
    return bonus;
}

bool Evaluate::is_isolated_pawn(const Board& board, Color color, int file)
{
    // check adjacent files for friendly pawns
    uint64_t adjacent_files = 0;
    if(file > 0)
        adjacent_files |= board.pieces(color, PieceType::Pawn) & (0x0101010101010101ULL << (file - 1));
    if(file < 7)
        adjacent_files |= board.pieces(color, PieceType::Pawn) & (0x0101010101010101ULL << (file + 1));
    
    return adjacent_files == 0;
}

bool Evaluate::is_doubled_pawn(const Board& board, Color color, int file)
{
    uint64_t file_pawns = board.pieces(color, PieceType::Pawn) & (0x0101010101010101ULL << file);
    
    return __builtin_popcountll(file_pawns) > 1;
}

bool Evaluate::is_backward_pawn(const Board& board, Color color, int square)
{
    int file = square & 7;
    int rank = square >> 3;
    
    // check if pawn can be supported by adjacent pawns
    bool can_be_supported = false;
    
    // left file
    if(file > 0)
    {
        uint64_t left_file_pawns = board.pieces(color, PieceType::Pawn) & (0x0101010101010101ULL << (file - 1));
        while(left_file_pawns)
        {
            int pawn_sq = __builtin_ctzll(left_file_pawns);

            int pawn_rank = pawn_sq >> 3;
            
            if(color == Color::White && pawn_rank <= rank)
                can_be_supported = true;
            if(color == Color::Black && pawn_rank >= rank)
                can_be_supported = true;
            
            left_file_pawns &= left_file_pawns - 1;
        }
    }
    
    // right file
    if(file < 7)
    {
        uint64_t right_file_pawns = board.pieces(color, PieceType::Pawn) & (0x0101010101010101ULL << (file + 1));
        while(right_file_pawns)
        {
            int pawn_sq = __builtin_ctzll(right_file_pawns);

            int pawn_rank = pawn_sq >> 3;
            
            if(color == Color::White && pawn_rank <= rank)
                can_be_supported = true;
            if(color == Color::Black && pawn_rank >= rank)
                can_be_supported = true;
            
            right_file_pawns &= right_file_pawns - 1;
        }
    }
    
    // check if square in front is controlled by enemy pawn
    bool blocked_by_enemy = false;
    int front_sq = (color == Color::White) ? square + 8 : square - 8;
    
    if(front_sq >= 0 && front_sq < 64)
    {
        uint64_t enemy_pawn_attacks = 0;
        if(color == Color::White)
        {
            // check for black pawn attacks
            if((front_sq & 7) > 0)
                enemy_pawn_attacks |= (1ULL << (front_sq + 7));
            if((front_sq & 7) < 7)
                enemy_pawn_attacks |= (1ULL << (front_sq + 9));
        } else
        {
            // check for white pawn attacks
            if((front_sq & 7) > 0)
                enemy_pawn_attacks |= (1ULL << (front_sq - 9));
            if((front_sq & 7) < 7)
                enemy_pawn_attacks |= (1ULL << (front_sq - 7));
        }
        
        blocked_by_enemy = (board.pieces(color == Color::White ? Color::Black : Color::White, PieceType::Pawn) & enemy_pawn_attacks) != 0;
    }
    
    return !can_be_supported && blocked_by_enemy;
}

bool Evaluate::is_passed_pawn(const Board& board, Color color, int square)
{
    int file = square & 7;
    int rank = square >> 3;
    
    // passed pawn zone: file and adjacent files ahead of the pawn
    uint64_t passed_zone = 0;
    
    if(color == Color::White)
    {
        // check ranks ahead
        for(int r = rank + 1; r < 8; r++)
        {
            if(file > 0)
                passed_zone |= (1ULL << (r * 8 + file - 1));

            passed_zone |= (1ULL << (r * 8 + file));

            if(file < 7)
                passed_zone |= (1ULL << (r * 8 + file + 1));
        }
    } else
    {
        // check ranks ahead
        for(int r = rank - 1; r >= 0; r--)
        {
            if(file > 0)
                passed_zone |= (1ULL << (r * 8 + file - 1));

            passed_zone |= (1ULL << (r * 8 + file));
            
            if(file < 7)
                passed_zone |= (1ULL << (r * 8 + file + 1));
        }
    }
    
    // check if any enemy pawns are in the passed zone
    uint64_t enemy_pawns = board.pieces(color == Color::White ? Color::Black : Color::White, PieceType::Pawn);
    
    return (enemy_pawns & passed_zone) == 0;
}

int Evaluate::passed_pawn_value(int rank, double phase)
{
    int base_value = Values::base_values[rank];
    
    // more valuable in endgame
    double endgame_multiplier = 1.0 + (1.0 - phase) * 1.5;
    
    return static_cast<int>(base_value * endgame_multiplier);
}

}

File: tuning/scripts/chess_evaluation/evaluate.py
from .types import EvaluationParams, ChessPosition

class ChessEvaluator:
    """Python implementation of the chess evaluation function"""
    
    def __init__(self, params: EvaluationParams = None):
        self.params = params or EvaluationParams()
        
        # phase weights for game phase calculation
        self.phase_weights = [0, 1, 1, 2, 4, 0] # P, N, B, R, Q, K
        self.max_phase = (1 * 2 + 1 * 2 + 2 * 2 + 4 * 1) * 2 # 24
        
        # opening PST
        self.pst_opening = {
            'P': [0, 0, 0, 0, 0, 0, 0, 0,
                  5, 10, 10, -20, -20, 10, 10, 5,
                  10, 10, 20, 30, 30, 20, 10, 10,
                  5, 5, 10, 50, 50, 10, 5, 5,
                  0, 0, 0, 60, 60, 0, 0, 0,
                  5, -5, -10, 20, 20, -10, -5, 5,
                  5, 10, 10, -20, -20, 10, 10, 5,
                  0, 0, 0, 0, 0, 0, 0, 0],
            
            'N': [-50, -40, -30, -30, -30, -30, -40, -50,
                  -40, -20, 0, 0, 0, 0, -20, -40,
                  -30, 0, 10, 15, 15, 10, 0, -30,
                  -30, 5, 15, 20, 20, 15, 5, -30,
                  -30, 0, 15, 20, 20, 15, 0, -30,
                  -30, 5, 10, 15, 15, 10, 5, -30,
                  -40, -20, 0, 5, 5, 0, -20, -40,
                  -50, -40, -30, -30, -30, -30, -40, -50],
            
            'B': [-20, -10, -10, -10, -10, -10, -10, -20,
                  -10, 0, 0, 0, 0, 0, 0, -10,
                  -10, 0, 5, 10, 10, 5, 0, -10,
                  -10, 5, 5, 10, 10, 5, 5, -10,
                  -10, 0, 10, 10, 10, 10, 0, -10,
                  -10, 10, 10, 10, 10, 10, 10, -10,
                  -10, 5, 0, 0, 0, 0, 5, -10,
                  -20, -10, -10, -10, -10, -10, -10, -20],
            
            'R': [0, 0, 0, 0, 0, 0, 0, 0,
                  -5, 0, 0, 0, 0, 0, 0, -5,
                  -5, 0, 0, 0, 0, 0, 0, -5,
                  -5, 0, 0, 0, 0, 0, 0, -5,
                  -5, 0, 0, 0, 0, 0, 0, -5,
                  -5, 0, 0, 0, 0, 0, 0, -5,
                  5, 10, 10, 10, 10, 10, 10, 5,
                  0, 0, 0, 5, 5, 0, 0, 0],
            
            'Q': [-20, -10, -10, -5, -5, -10, -10, -20,
                  -10, 0, 0, 0, 0, 0, 0, -10,
                  -10, 0, 5, 5, 5, 5, 0, -10,
                  -5, 0, 5, 5, 5, 5, 0, -5,
                  0, 0, 5, 5, 5, 5, 0, -5,
                  -10, 5, 5, 5, 5, 5, 0, -10,
                  -10, 0, 5, 0, 0, 0, 0, -10,
                  -20, -10, -10, -5, -5, -10, -10, -20],
            
            'K': [20, 30, 10, 0, 0, 10, 30, 20,
                  20, 20, 0, 0, 0, 0, 20, 20,
                  -10, -20, -20, -20, -20, -20, -20, -10,
                  -20, -30, -30, -40, -40, -30, -30, -20,
                  -30, -40, -40, -50, -50, -40, -40, -30,
                  -30, -40, -40, -50, -50, -40, -40, -30,
                  -30, -40, -40, -50, -50, -40, -40, -30,
                  -30, -40, -40, -50, -50, -40, -40, -30]
        }
        
        # endgame PST
        self.pst_endgame = {
            'P': [0, 0, 0, 0, 0, 0, 0, 0,
                  5, 10, 10, -20, -20, 10, 10, 5,
                  5, -5, -10, 0, 0, -10, -5, 5,
                  0, 0, 0, 20, 20, 0, 0, 0,
                  5, 5, 10, 25, 25, 10, 5, 5,
                  10, 10, 20, 30, 30, 20, 10, 10,
                  50, 50, 50, 50, 50, 50, 50, 50,
                  0, 0, 0, 0, 0, 0, 0, 0],
            
            'N': [-50, -40, -30, -30, -30, -30, -40, -50,
                  -40, -20, 0, 0, 0, 0, -20, -40,
                  -30, 0, 10, 15, 15, 10, 0, -30,
                  -30, 5, 15, 20, 20, 15, 5, -30,
                  -30, 0, 15, 20, 20, 15, 0, -30,
                  -30, 5, 10, 15, 15, 10, 5, -30,
                  -40, -20, 0, 5, 5, 0, -20, -40,
                  -50, -40, -30, -30, -30, -30, -40, -50],
            
            'B': [-20, -10, -10, -10, -10, -10, -10, -20,
                  -10, 0, 0, 0, 0, 0, 0, -10,
                  -10, 0, 5, 10, 10, 5, 0, -10,
                  -10, 5, 5, 10, 10, 5, 5, -10,
                  -10, 0, 10, 10, 10, 10, 0, -10,
                  -10, 10, 10, 10, 10, 10, 10, -10,
                  -10, 5, 0, 0, 0, 0, 5, -10,
                  -20, -10, -10, -10, -10, -10, -10, -20],
            
            'R': [0, 0, 5, 10, 10, 5, 0, 0,
                  0, 0, 5, 10, 10, 5, 0, 0,
                  0, 0, 5, 10, 10, 5, 0, 0,
                  0, 0, 5, 10, 10, 5, 0, 0,
                  0, 0, 5, 10, 10, 5, 0, 0,
                  0, 0, 5, 10, 10, 5, 0, 0,
                  25, 25, 25, 25, 25, 25, 25, 25,
                  0, 0, 5, 10, 10, 5, 0, 0],
            
            'Q': [-20, -10, -10, -5, -5, -10, -10, -20,
                  -10, 0, 0, 0, 0, 0, 0, -10,
                  -10, 0, 5, 5, 5, 5, 0, -10,
                  -5, 0, 5, 5, 5, 5, 0, -5,
                  0, 0, 5, 5, 5, 5, 0, -5,
                  -10, 5, 5, 5, 5, 5, 0, -10,
                  -10, 0, 5, 0, 0, 0, 0, -10,
                  -20, -10, -10, -5, -5, -10, -10, -20],
            
            'K': [-50, -40, -30, -20, -20, -30, -40, -50,
                  -30, -20, -10, 0, 0, -10, -20, -30,
                  -30, -10, 20, 30, 30, 20, -10, -30,
                  -30, -10, 30, 40, 40, 30, -10, -30,
                  -30, -10, 30, 40, 40, 30, -10, -30,
                  -30, -10, 20, 30, 30, 20, -10, -30,
                  -30, -30, 0, 0, 0, 0, -30, -30,
                  -50, -30, -30, -30, -30, -30, -30, -50]
        }

    def evaluate(self, position: ChessPosition) -> float:
        """Main evaluation function"""

        score = 0.0
        
        phase = self._calculate_phase(position)
        
        score += self._evaluate_material(position, phase)
        score += self._evaluate_castling(position, phase)
        score += self._evaluate_pawn_structure(position, phase)
        
        # return score from white's perspective
        return score if position.turn == 'w' else -score

    def _calculate_phase(self, position: ChessPosition) -> float:
        """Calculate game phase (1.0 = opening, 0.0 = endgame)"""

        phase = 0
        
        piece_types = ['P', 'N', 'B', 'R', 'Q']
        for i, piece_type in enumerate(piece_types):
            if i == 0: # skip pawns for phase calculation
                continue
            
            count = (position.get_piece_count('w', piece_type) + 
                    position.get_piece_count('b', piece_type))
            phase += self.phase_weights[i] * count
        
        if phase < 0:
            phase = 0
        
        return min(phase / self.max_phase, 1.0)
    
    def _evaluate_material(self, position: ChessPosition, phase: float) -> float:
        """Evaluate material balance and piece-square tables"""

        score = 0.0
        opening_score = 0.0
        endgame_score = 0.0
        
        piece_types = ['P', 'N', 'B', 'R', 'Q', 'K']
        
        for i, piece_type in enumerate(piece_types):
            # white pieces
            for square in position.white_pieces[piece_type]:
                score += self.params.material_value[i]
                opening_score += self.pst_opening[piece_type][square]
                endgame_score += self.pst_endgame[piece_type][square]
            
            # white pieces (mirror vertically)
            for square in position.black_pieces[piece_type]:
                score -= self.params.material_value[i]
                mirrored_square = square ^ 56  # vertical flip
                opening_score -= self.pst_opening[piece_type][mirrored_square]
                endgame_score -= self.pst_endgame[piece_type][mirrored_square]
        
        # blend opening and endgame scores
        blended_pst = opening_score * phase + endgame_score * (1.0 - phase)
        
        return score + blended_pst
    
    def _evaluate_castling(self, position: ChessPosition, phase: float) -> float:
        """Evaluate castling rights and castled positions"""

        score = 0.0
        
        # castling rights bonus
        if 'K' in position.castling or 'Q' in position.castling:
            score += self.params.castle_rights_bonus
        if 'k' in position.castling or 'q' in position.castling:
            score -= self.params.castle_rights_bonus
        
        # castled position detection
        if position.white_pieces['K']:
            wk_square = position.white_pieces['K'][0]
            if wk_square == 6 or wk_square == 2: # g1 or c1
                score += self.params.castled_position_bonus
        
        if position.black_pieces['K']:
            bk_square = position.black_pieces['K'][0]
            if bk_square == 62 or bk_square == 58: # g8 or c8
                score -= self.params.castled_position_bonus
        
        # blend by opening phase
        return score * phase
    
    def _evaluate_pawn_structure(self, position: ChessPosition, phase: float) -> float:
        """Evaluate pawn structure"""

        score = 0.0
        
        # white pawns
        score += self._analyze_pawn_weaknesses(position, 'w', phase)
        score += self._analyze_passed_pawns(position, 'w', phase)
        
        # black pawns
        score -= self._analyze_pawn_weaknesses(position, 'b', phase)
        score -= self._analyze_passed_pawns(position, 'b', phase)
        
        return score
    
    def _analyze_pawn_weaknesses(self, position: ChessPosition, color: str, phase: float) -> float:
        """Analyze pawn weaknesses for given color"""

        penalty = 0.0
        
        pawns = position.white_pieces['P'] if color == 'w' else position.black_pieces['P']
        
        # count pawns per file
        file_counts = [0] * 8
        for square in pawns:
            file_counts[square % 8] += 1
        
        for square in pawns:
            file = square % 8
            
            # isolated pawn penalty
            if self._is_isolated_pawn(position, color, file):
                isolated_penalty = self.params.isolated_pawn_penalty
                # worse in endgame
                isolated_penalty *= (1.0 + (1.0 - phase) * 0.5)
                penalty += isolated_penalty
            
            # doubled pawn penalty
            if file_counts[file] > 1:
                penalty += self.params.doubled_pawn_penalty
            
            # backward pawn penalty
            if self._is_backward_pawn(position, color, square):
                penalty += self.params.backward_pawn_penalty
        
        return -penalty
    
    def _analyze_passed_pawns(self, position: ChessPosition, color: str, phase: float) -> float:
        """Analyze passed pawns for given color"""

        bonus = 0.0
        
        pawns = position.white_pieces['P'] if color == 'w' else position.black_pieces['P']
        
        for square in pawns:
            if self._is_passed_pawn(position, color, square):
                rank = (square // 8) if color == 'w' else (7 - (square // 8))
                bonus += self._passed_pawn_value(rank, phase)
                
                # additional bonuses for advanced passed pawns
                if rank >= 5:
                    file = square % 8
                    
                    # connected passed pawn bonus
                    if self._has_adjacent_pawns(position, color, file):
                        bonus += self.params.connected_passed_pawn_bonus
                    
                    # protected passed pawn bonus
                    if self._is_protected_by_pawn(position, color, square):
                        bonus += self.params.protected_passed_pawn_bonus
        
        return bonus
    
    def _is_isolated_pawn(self, position: ChessPosition, color: str, file: int) -> bool:
        """Check if pawn on given file is isolated"""

        pawns = position.white_pieces['P'] if color == 'w' else position.black_pieces['P']
        
        # check adjacent files
        for square in pawns:
            pawn_file = square % 8
            if abs(pawn_file - file) == 1:
                return False
        
        return True
    
    def _is_backward_pawn(self, position: ChessPosition, color: str, square: int) -> bool:
        """Check if pawn is backward"""

        file = square % 8
        rank = square // 8
        
        own_pawns = position.white_pieces['P'] if color == 'w' else position.black_pieces['P']
        enemy_pawns = position.black_pieces['P'] if color == 'w' else position.white_pieces['P']
        
        # check if can be supported by adjacent pawns
        can_be_supported = False
        for adj_file in [file - 1, file + 1]:
            if 0 <= adj_file < 8:
                for pawn_square in own_pawns:
                    if pawn_square % 8 == adj_file:
                        pawn_rank = pawn_square // 8
                        if (color == 'w' and pawn_rank <= rank) or (color == 'b' and pawn_rank >= rank):
                            can_be_supported = True
                            break
        
        # check if blocked by enemy pawn
        front_square = square + 8 if color == 'w' else square - 8
        if not (0 <= front_square < 64):
            return False
        
        blocked_by_enemy = False
        for enemy_square in enemy_pawns:
            enemy_file = enemy_square % 8
            enemy_rank = enemy_square // 8
            
            # check if enemy pawn attacks the front square
            if abs(enemy_file - (front_square % 8)) == 1:
                if (color == 'w' and enemy_rank == front_square // 8 + 1) or (color == 'b' and enemy_rank == front_square // 8 - 1):
                    blocked_by_enemy = True
                    break
        
        return not can_be_supported and blocked_by_enemy
    
    def _is_passed_pawn(self, position: ChessPosition, color: str, square: int) -> bool:
        """Check if pawn is passed"""

        file = square % 8
        rank = square // 8
        
        enemy_pawns = position.black_pieces['P'] if color == 'w' else position.white_pieces['P']
        
        # check if any enemy pawns can stop this pawn
        for enemy_square in enemy_pawns:
            enemy_file = enemy_square % 8
            enemy_rank = enemy_square // 8
            
            # check if enemy pawn is in the passed pawn zone
            if abs(enemy_file - file) <= 1:
                if (color == 'w' and enemy_rank > rank) or (color == 'b' and enemy_rank < rank):
                    return False
        
        return True
    
    def _passed_pawn_value(self, rank: int, phase: float) -> float:
        """Calculate value of passed pawn based on rank"""

        base_value = self.params.base_values[rank]
        
        # more valuable in endgame
        endgame_multiplier = 1.0 + (1.0 - phase) * 1.5
        
        return base_value * endgame_multiplier
    
    def _has_adjacent_pawns(self, position: ChessPosition, color: str, file: int) -> bool:
        """Check if there are adjacent pawns"""

        return False # placeholder
    
    def _is_protected_by_pawn(self, position: ChessPosition, color: str, square: int) -> bool:
        """Check if pawn is protected by another pawn"""

        return False # placeholder